### PR TITLE
 vmss create: more scale out support

### LINKED ISSUE #5920: vmss create: support to configure FD count
//CC: @gatneil. I thought I could get it done w/o a tracking issue, but so far I can't get the right payload submitted to make the CRP happy. So opening this issue as it might take come time to finalize.

```
If the customer chooses only 1 fault domain, they cannot choose single placement group; they must use multiple placement groups. For CLI, 2 changes are required:

1)	If the customer chooses 1 fault domain in the CLI but doesn’t specify placement group, create the scale set with multiple placement groups (note: in zones, the service-level default of multiple placement groups should make this fine, but for non-zonal scale sets, the service-level default is still single placement groups, which will cause an error).
2)	If the customer chooses 1 fault domain and single placement group in the CLI, then error out with “With platform fault domain count 1, you must use multiple placement groups”.
```

### LINKED ISSUE #5921: vmss create: default to Standard LB if conditions are met
//CC @gatneil, @christiankuhtz 
Any one of of the following will cause defaults to switch to `standard`
1. Multiple group placement is on
2.  Zonal 
3. Instance-count > 100

### PR DESCRIPTION
**Pull request by yugangw-msft**
2 works (Fix #5920, Fix #5921)
* `vmss create`: support to configure platform fault domain count through `--platform-fault-domain-count`
* `vmss create`: default to Standard LB for zonal, large, or single-placement-group disabled scale-set

@gatneil, for unconditionally defaulting to `Standard` LB, i will do it later as that would be a breaking change and we can't roll out right now.
//CC @christiankuhtz 

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).
- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5987
This is an **experimental preview** for @microsoft users.

**Comment by gatneil** [comment]
Cool. Thanks, @yugangw-msft! :)

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py [hunk 1]
```diff
@@ -265,7 +265,8 @@ def load_arguments(self, _):
         VMPriorityTypes = self.get_models('VirtualMachinePriorityTypes', resource_type=ResourceType.MGMT_COMPUTE)
         c.argument('name', name_arg_type)
         c.argument('nat_backend_port', default=None, help='Backend port to open with NAT rules.  Defaults to 22 on Linux and 3389 on Windows.')
-        c.argument('single_placement_group', default=None, help="Enable single placement group. This flag will default to True if instance count <=100, and default to False for instance count >100.", arg_type=get_enum_type(['true', 'false']))
+        c.argument('single_placement_group', default=None, help="Enable replicate using fault domains within the same cluster", arg_type=get_enum_type(['true', 'false']))
+        c.argument('platform_fault_domain_count', default=None, type=int, help='Fault Domain count for each placement group in the availability zone', min_api='2017-12-01')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py]: Why are you specifying `default=None`?  VMSS create is a custom command, so that should be established in its signature.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py [hunk 2]
```diff
@@ -285,7 +286,8 @@ def load_arguments(self, _):
...
         c.argument('backend_port', help='When creating a new load balancer, backend port to open with NAT rules (Defaults to 22 on Linux and 3389 on Windows). When creating an application gateway, the backend port to use for the backend HTTP settings.', type=int)
         c.argument('load_balancer', help='Name to use when creating a new load balancer (default) or referencing an existing one. Can also reference an existing load balancer by ID or specify "" for none.', options_list=['--load-balancer', '--lb'])
-        c.argument('load_balancer_sku', resource_type=ResourceType.MGMT_NETWORK, min_api='2017-08-01', help="SKU when creating a new Load Balancer. Default to 'Basic' for any non-zonal scaleset and 'Standard' otherwise", options_list=['--lb-sku'], arg_type=get_enum_type(LoadBalancerSkuName))
+        c.argument('load_balancer_sku', resource_type=ResourceType.MGMT_NETWORK, min_api='2017-08-01', options_list=['--lb-sku'], arg_type=get_enum_type(LoadBalancerSkuName),
+                   help="Sku of the Load Balancer to create. Default to 'Standard' for zonal or big scale-set, and 'Basic' otherwise")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py]: SKU.  Also, what constitutes "big scale-set"?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 3]
```diff
@@ -943,41 +939,40 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
     return values[0]
 
 
-def _validate_vmss_create_load_balancer_or_app_gateway(cmd, namespace):
-    from msrestazure.azure_exceptions import CloudError
-    from msrestazure.tools import parse_resource_id
-    INSTANCE_THRESHOLD = _get_vmss_create_instance_threshold()
+def _validate_vmss_single_placement_group(namespace):
+    if namespace.platform_fault_domain_count is not None and namespace.zones is None:
+        raise CLIError('usage error: --platform-fault-domain-count <COUNT> --zones <ZONES>')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: The usage statement shouldn't have the angle brackets, just `COUNT` and `ZONES` respectively.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 4]
```diff
@@ -943,41 +939,40 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
...
+    if namespace.platform_fault_domain_count == 1 or namespace.zones or namespace.instance_count > 100:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: Why did you remove the method that retrieves this value?  There was a test for this scenario that needed to be able to mock the threshold to something lower (since I believe creating the 100+ instance scale set actually just couldn't be done on our subscription due to quota limits).

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 5]
```diff
@@ -943,41 +939,40 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
...
+        elif namespace.single_placement_group == 'true':
+            raise CLIError("usage error: '--single-placement-group' should be turned off when one of the conditions"
+                           " are met: 1. Zonal, 2. 100+ instances, 3. fault domain count set to 1")
 
     # convert the single_placement_group to boolean for simpler logic beyond
-    if namespace.single_placement_group is None:
-        namespace.single_placement_group = namespace.instance_count <= INSTANCE_THRESHOLD
-    else:
-        namespace.single_placement_group = (namespace.single_placement_group == 'true')
+    namespace.single_placement_group = (namespace.single_placement_group != 'false')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: You should be able to register the argument to get actual Boolean values instead of strings 'true' and 'false'.  That may simplify your code (or not!).

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 6]
```diff
@@ -943,41 +939,40 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
...
-        logger.debug("defaulting to '%s' because instance count <= %s", balancer_type, INSTANCE_THRESHOLD)
-        if balancer_type == 'applicationGateway':
-            logger.warning("In a future release of CLI, for scalesets with 100+ instances, the default balancer "
-                           "will be Standard Load Balancer instead of Application Gateway. Use '--app-gateway' to "
-                           "continue using applicaiton gateways")
+        if std_lb_is_available:
+            balancer_type = 'loadBalancer'
+        else:
+            balancer_type = 'loadBalancer' if namespace.single_placement_group else 'applicationGateway'
+            logger.debug("W/o STD LB, defaulting to '%s' under because single placement group is disabled",
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: So if the profile doesn't support standard LB, it uses the old rules?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 7]
```diff
@@ -1052,6 +1047,14 @@ def _validate_vmss_create_load_balancer_or_app_gateway(cmd, namespace):
             namespace.load_balancer_type = 'new'
             logger.debug('new load balancer will be created')
 
+        if namespace.load_balancer_type == 'new' and not namespace.single_placement_group and std_lb_is_available:
+            LBSkuName = cmd.get_models('LoadBalancerSkuName', resource_type=ResourceType.MGMT_NETWORK)
+            if namespace.load_balancer_sku is None:
+                namespace.load_balancer_sku = LBSkuName.standard.value
+                logger.debug("defaulting to Standard Load Balancer for the big scale-set")
+            elif namespace.load_balancer_sku == LBSkuName.basic.value:
+                raise CLIError("usage error: Standard load balancer is required based on other arguments provided")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: Which arguments?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 8]
```diff
@@ -1052,6 +1047,14 @@ def _validate_vmss_create_load_balancer_or_app_gateway(cmd, namespace):
             namespace.load_balancer_type = 'new'
             logger.debug('new load balancer will be created')
 
+        if namespace.load_balancer_type == 'new' and not namespace.single_placement_group and std_lb_is_available:
+            LBSkuName = cmd.get_models('LoadBalancerSkuName', resource_type=ResourceType.MGMT_NETWORK)
+            if namespace.load_balancer_sku is None:
+                namespace.load_balancer_sku = LBSkuName.standard.value
+                logger.debug("defaulting to Standard Load Balancer for the big scale-set")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: "Default to Standard Load Balancer for scale-set larger than <VALUE>".  What does "big scale-set" mean?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py [hunk 9]
```diff
@@ -1939,14 +1943,6 @@ def _get_public_ip_address_allocation(value, sku):
     if secrets:
         secrets = _merge_secrets([validate_file_or_dict(secret) for secret in secrets])
 
-    if nsg is None and zones:
-        # Per https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-standard-overview#nsg
-        nsg_name = '{}NSG'.format(vmss_name)
-        master_template.add_resource(build_nsg_resource(
-            None, nsg_name, location, tags, 'rdp' if os_type.lower() == 'windows' else 'ssh'))
-        nsg = "[resourceId('Microsoft.Network/networkSecurityGroups', '{}')]".format(nsg_name)
-        vmss_dependencies.append('Microsoft.Network/networkSecurityGroups/{}'.format(nsg_name))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: Did this move somewhere?  If not, what is being removed and what impact does that have?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 10]
```diff
@@ -1505,6 +1505,23 @@ def test_vmss_existing_lb(self, resource_group):
         self.cmd('network lb create -g {rg} -n {lb} --backend-pool-name test')
         self.cmd('vmss create -g {rg} -n {vmss} --load-balancer {lb} --image UbuntuLTS --admin-username clitester --admin-password TestTest12#$')
 
+    @ResourceGroupPreparer()
+    def test_vmss_single_placement_group_default_to_std_lb(self, resource_group):
+        self.kwargs.update({
+            'vmss': 'vmss123'
+        })
+        self.cmd('vmss create -g {rg} -n {vmss} --admin-username clitester --admin-password PasswordPassword1! --image debian --single-placement-group false')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: The logic is getting very complex for scale sets. I'd recommend putting a comment here on what the expected behavior is and why,

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py (hunk 4 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: Can I leverage the `--validate` flag and check out the template?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 11]
```diff
@@ -943,41 +939,37 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
     return values[0]
 
 
+def _validate_vmss_single_placement_group(namespace):
+    if namespace.platform_fault_domain_count is not None and namespace.zones is None:
+        raise CLIError('usage error: --platform-fault-domain-count COUNT --zones ZONES')
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: @gatneil, I am adding this client side validation to ensure users can only set `platformFaultDomainCount` on zonal scale-sets, as the error from service side is rather vague. However if CRP is getting it work with non-zonal ones soon, then I don't think we need such validation as it would end up blocking people. Thoughts?

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py [hunk 12]
```diff
@@ -265,8 +265,9 @@ def load_arguments(self, _):
         VMPriorityTypes = self.get_models('VirtualMachinePriorityTypes', resource_type=ResourceType.MGMT_COMPUTE)
         c.argument('name', name_arg_type)
         c.argument('nat_backend_port', default=None, help='Backend port to open with NAT rules.  Defaults to 22 on Linux and 3389 on Windows.')
-        c.argument('single_placement_group', default=None, help="Enable replicate using fault domains within the same cluster", arg_type=get_enum_type(['true', 'false']))
-        c.argument('platform_fault_domain_count', default=None, type=int, help='Fault Domain count for each placement group in the availability zone', min_api='2017-12-01')
+        c.argument('single_placement_group', arg_type=get_three_state_flag(), help="Enable replicate using fault domains within the same cluster. Default to 'false' for any zonals, or with 100+ instances, or Fault Domain count set at 1"
+                   " See https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-placement-groups for details")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_params.py]: This should really be in the long-summary for this param, so this would need to be in the `help.py` file.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py (hunk 4 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: That might work, though I think it outputs the template as log.info so it might be tricky to get to.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 13]
```diff
@@ -941,16 +941,13 @@ def _get_default_address_pool(cli_ctx, resource_group, balancer_name, balancer_t
...
 def _validate_vmss_single_placement_group(namespace):
     if namespace.platform_fault_domain_count is not None and namespace.zones is None:
-        raise CLIError('usage error: --platform-fault-domain-count <COUNT> --zones <ZONES>')
+        raise CLIError('usage error: --platform-fault-domain-count COUNT --zones ZONES')
     if namespace.platform_fault_domain_count == 1 or namespace.zones or namespace.instance_count > 100:
         if namespace.single_placement_group is None:
-            namespace.single_placement_group = 'false'
-        elif namespace.single_placement_group == 'true':
+            namespace.single_placement_group = False
+        elif namespace.single_placement_group is True:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: This can just be `elif namespace.single_placement_group:`

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 14]
```diff
@@ -1047,13 +1044,15 @@ def _validate_vmss_create_load_balancer_or_app_gateway(cmd, namespace):
...
             LBSkuName = cmd.get_models('LoadBalancerSkuName', resource_type=ResourceType.MGMT_NETWORK)
             if namespace.load_balancer_sku is None:
                 namespace.load_balancer_sku = LBSkuName.standard.value
-                logger.debug("defaulting to Standard Load Balancer for the big scale-set")
+                logger.debug("use Standard sku as single placement group is turned off")
             elif namespace.load_balancer_sku == LBSkuName.basic.value:
-                raise CLIError("usage error: Standard load balancer is required based on other arguments provided")
+                raise CLIError("usage error: 'Standard' load balancer is required because 'single placement group' is "
+                               "turned off to accomodate zonal scale-sets, or with 100+ instances, or Fault Domain "
+                               "count is set to 1")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: It would be better from a user perspective to spell out which of these is the culprit, but this is better than the previous error for sure.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py [hunk 15]
```diff
@@ -968,8 +965,8 @@ def _validate_vmss_create_load_balancer_or_app_gateway(cmd, namespace):
     if namespace.load_balancer is None and namespace.application_gateway is None:
         if std_lb_is_available:
             balancer_type = 'loadBalancer'
-        else:
-            balancer_type = 'loadBalancer' if namespace.single_placement_group else 'applicationGateway'
+        else:  # mainly for Azure stack users
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py]: Might be better to make this comment stronger. 
`# needed for Stack profile 2017_03_09`
